### PR TITLE
Allow device to fill in requested permissions in token request to server

### PR DESCRIPTION
**Pull request by JohnySeven**
Hi Guys,

I've implemented sending permissions in token request to SignalK server - tested with readwrite (default) permission and also with admin permission. Solves issue #173 

I've changed set_sk_server method on Builder and added third parameter permission with default READWRITE value. I've have changed SensESPApp ctor paramteters to add permision. Same with WSClient.

Let me know what you think!
Jan

### CODE REVIEW

File: src/sensesp_app.cpp [hunk 1]
```diff
@@ -169,6 +171,26 @@ void SensESPApp::reset() {
   app.onDelay(1000, []() { ESP.restart(); });
 }
 
+String SensESPApp::get_permission_string(SKPermissions permission) 
+{
+  if(permission == READONLY)
```
Comment (mairas) [inline src/sensesp_app.cpp:176]: You can achieve the same neater (and arguably more efficiently) by placing the permission strings in an array of strings and then referencing the array by coercing the enum to an int:

```
char* permission_strings[] = {
  "readonly",
  "readwrite",
  "admin"
};

...

permission_string = permission_strings[(int)permission];
```

(Not sure if the type definitions are correct in my example, but you get the point. Also, didn't check if the enums were in that order.)
};

**Comment by mairas** [comment]
This looks golden, except for the small comment/suggestion I made. And seems like it's now conflicting with some prior changes. If you need help with resolving the conflict, send me a message.

**Comment by ba58smith** [comment]
What is the default expiration of permissions if they are set this way? Hopefully "NEVER".

**Comment by mairas** [comment]
> What is the default expiration of permissions if they are set this way? Hopefully "NEVER".

Blank, I suppose, if the server didn't support setting the expiration for auth requests. But possibly a better fix would be to set the default expiration on the server side to NEVER? @tkurki

File: src/sensesp_app.cpp (hunk 1 above)
Comment (JohnySeven) [reply src/sensesp_app.cpp:176]: Good point, will change that!

**Comment by JohnySeven** [comment]
Yes, it's blank and it would be nice if the server would fill in default from configuration - "NEVER" would be fine. I hope somebody from node server team will maybe do it.

**Comment by ba58smith** [comment]
Does it seem a bit risky to implement this change without knowing how the server is going to handle a blank expiration?

**Comment by mairas** [comment]
> Does it seem a bit risky to implement this change without knowing how the server is going to handle a blank expiration?

This is only about an auth request. The user/admin still needs to go and approve the request. The only difference is that now the required permissions are pre-filled.

**Comment by ba58smith** [comment]
> > Does it seem a bit risky to implement this change without knowing how the server is going to handle a blank expiration?
> 
> This is only about an auth request. The user/admin still needs to go and approve the request. The only difference is that now the required permissions are pre-filled.

Ah, I see. Just trying to prevent someone from approving a request for "ReadOnly" because they don't change it to "ReadWrite". Got it.

**Comment by mairas** [comment]
It wouldn't be much of an authentication scheme if the client was allowed to authorize itself! :-D

**Comment by ba58smith** [comment]
> It wouldn't be much of an authentication scheme if the client was allowed to authorize itself! :-D

Oh, you and your LOGIC! :-)

File: src/sensesp_app.cpp (hunk 1 above)
Comment (JohnySeven) [reply src/sensesp_app.cpp]: Done

**Comment by JohnySeven** [comment]
@mairas it's ready to be merged, thanks!

**Comment by mairas** [comment]
Before I merge, could you remove the last merge commit from the PR?

**Comment by JohnySeven** [comment]
I've reverted merge commit from repo. Thanks

**Comment by tedenda** [comment]
My guess is that something went wrong?
Now there is 153 files changed and conflicts?

**Comment by mairas** [comment]
This PR indeed has quite a few extra commits and dozens of unrelated files on it. Instead of adding revert commits (which just add more stuff on the repo), remove the extra commits, either with `git reset --hard [...]` or `git rebase -i [...]`. Then rebase your branch against the latest master, and once that's done, force-push the changes to your forked repo. That'll fix the PR.

**Comment by tedenda** [comment]
Fixes #173 when merged

**Comment by JohnySeven** [comment]
I've done what @mairas suggested and PR is now in good shape! Thanks @mairas :-)

**Comment by mairas** [comment]
I'm sorry for this to come in so late, but now that I began thinking about it, is there ever a reason for requesting any other than readwrite permission? Even if you implemented a pure control device, you'd still want to report back the state of the controlled interface, so readonly is not feasible. And on the other hand, "admin" is not relevant either because I don't see SensESP devices mucking with the server in such a manner.

Assuming this, I'd simply hard-code the value and remove all that complexity-adding configurability.

**Comment by JohnySeven** [comment]
@mairas readonly can be used for pure displays that just show boat information, admin can be used for sending put requests to control another device on boat - I know it's not implemented right now, but it's something I would like to implement in future as the SK protocol allows this scenario and is used in KIP web app. The default value is readwrite and I don't think it's not a big change in code, nor some breaking change, it just adds more features to SensESP.

**Comment by mairas** [comment]
OK, you've got a point there. Merging!

**Comment by JohnySeven** [comment]
@mairas  thanks! :-)

**Pull request merged**